source/dklib/khashl.d

```diff
@@ -0,0 +1,565 @@
+    }
+
+    auto kh_capacity(const(kh_t)* h)
+    {
+        return h.keys ? 1U<<h.bits : 0U;
+    }
+
+}
+
+/** --- BEGIN OF HASH FUNCTIONS --- */
+template kh_hash(T)
+{
+pragma(inline, true)
+{
+    auto kh_hash_func(T)(T key)
+    if (is(T == uint) || is(T == uint32_t) || is(T == khint32_t))
+    {
+        key += ~(key << 15);
+        key ^=  (key >> 10);
+        key +=  (key << 3);
+        key ^=  (key >> 6);
+        key += ~(key << 11);
+        key ^=  (key >> 16);
+        return key;
+    }
+
+    auto kh_hash_func(T)(T key)
+    if (is(T == ulong) || is(T == uint64_t) || is(T == khint64_t))
+    {
+        key = ~key + (key << 21);
+        key = key ^ key >> 24;
+        key = (key + (key << 3)) + (key << 8);
+        key = key ^ key >> 14;
+        key = (key + (key << 2)) + (key << 4);
+        key = key ^ key >> 28;
+        key = key + (key << 31);
+        return cast(khint_t) key;
+    }
+
+    khint_t kh_hash_str(const(char)* s)
+    {
+        khint_t h = cast(khint_t)*s;
+        if (h) for  (++s; *s; ++s) h = (h << 5) - h + cast(khint_t)*s;
+        return h;
+    }
+    
+    auto kh_hash_func(T)(T* key)
+    if(is(T == char) || is(T == const(char)) || is(T == immutable(char)))
+    {
+        return kh_hash_str(key);
+    }
+
+    auto kh_hash_func(T)(T key)
+    if(isSomeString!T)
+    {
+        // rewrite __ac_X31_hash_string for D string/smart array
+        if (key.length == 0) return 0;
+        khint_t h = key[0];
+        for (int i=1; i<key.length; ++i)
+            h = (h << 5) - h + cast(khint_t) key[i];
+        return h;
+    }
+    
+    
+
+} // end pragma(inline, true)
+} // end template kh_hash
+
+/// In order to take advantage of cached-hashes
+/// our equality function will actually take the bucket type as opposed to just the key.
+/// This allows it to access both the store hash and the key itself.
+template kh_equal(T, bool cached)
+{
+pragma(inline,true)
+{
+    /// Assert that we are using a bucket type with key member
+    static assert(hasMember!(T, "key"));
+
+    /// Assert that we are using a bucket type with hash member if using hash-caching
+    static if(cached) static assert(hasMember!(T, "hash"));
+
+    bool kh_hash_equal(T)(T a, T b)
+    if (isNumeric!(typeof(__traits(getMember,T,"key"))))
+    {
+        /// There is no benefit to caching hashes for integer keys (I think)
+        static assert (cached == false, "No reason to cache hash for integer keys");
+        return (a.key == b.key);
+    }
+    
+    bool kh_hash_equal(T)(T* a, T* b)
+    if(
+        is(typeof(__traits(getMember,T,"key")) == char) || 
+        is(typeof(__traits(getMember,T,"key")) == const(char)) || 
+        is(typeof(__traits(getMember,T,"key")) == immutable(char)))
+    {
+        /// If using hash-caching we check equality of the hashes first 
+        /// before checking the equality of keys themselves 
+        static if(cached) return (a.hash == b.hash) && (strcmp(a, b) == 0);
```

Comment:
If `a` is the bucket type does this [`strcmp(a,b)`] really work? Need `char *` tests , see above